### DISCUSSION TITLE
CircleCI Cache strategy problem

### DISCUSSION THREAD

**Discussion by chenghuzi** [jonescompneurolab/hnn-core#591] [General]
It seems that, on the CircleCI platform, if you create a cached dir with a key, say, "key-A" in the first run, then no matter what you do with that directory, including overwriting its contents, removing sub-directories, the cached content won't change actually. That means the next time when you extract the cache with "key-A", you'll always get the content that you wrote the first time.

Here is an simple example:
[demo repo](https://github.com/chenghuzi/test_ci_cache_flush)

1. Create cache `examples_built ` with a random matrix: https://app.circleci.com/pipelines/github/chenghuzi/test_ci_cache_flush/3/workflows/1ab9e6d7-b00e-42fb-9032-cb0cc1c7628a/jobs/3
2. Remove the cached content with a `rm -rf` command and write something new into the dir: https://app.circleci.com/pipelines/github/chenghuzi/test_ci_cache_flush/4/workflows/5e5b5334-33e9-4b38-a725-53f46dc6548a/jobs/4/parallel-runs/0/steps/0-107
3. Try to extract the content to see if the cached content is updated: https://app.circleci.com/pipelines/github/chenghuzi/test_ci_cache_flush/5/workflows/8f30d40a-0951-4927-bad2-66596d116511/jobs/5/parallel-runs/0/steps/0-107

We can see that the matrix written in the first step starts with `0.87151745`, and the recreated matrix in step 2 starts with `0.85983841`. However, in step 3 we still get `0.87151745`. This shows that CircleCI does not allow any modifications to an existing cache, it looks weird to me though.

**Comment by jasmainak** [comment]
Thinking about it, it's not that weird. The cache key in many situations reflects the data that is stored in the cache. There should be a one-to-one mapping between the key and the data, not one to many ...

**Reply by chenghuzi** [reply]
yeah I understand in that sense the system is "pure", but that also makes it impossible for us to flush the build content. On the other hand, I'm curious about how they deal with the ever-increasing cache storage as the cache is also undeletable.

**Comment by jasmainak** [comment]
Here it says caches are immutable: https://circleci.com/blog/config-best-practices-dependency-caching/
